### PR TITLE
Bedrock support for C# and support for powershell version 5.1

### PR DESCRIPTION
**Pull request by Ajoro**
## Proposed Changes
- C#: Added stripped_motd support [See issue 84](https://github.com/FragLand/minestat/issues/84)
- C#: Added Bedrock request and parse [See issue 45](https://github.com/FragLand/minestat/issues/45)
- Updated C# example
- PS: Added function to convert json to hashtabel. Caused json requests to crash.
- PS: Removed "§" and added unicode reference, would cause some motds to parse incorrectly.
- PS: Added support for splitting address and port if address input is `Address:Port`
- Merged with https://github.com/FragLand/minestat/pull/107

### CODE REVIEW

**Comment by ldilley** [comment]
Apologies for my recent collision (PR #107) with your PowerShell changes in this PR, @Ajoro. I did not realize that this PR also touched those files when the notification came through in Discord (I only saw the C# portion). We can figure out how to merge your work. At any rate, thank you again for the incredible contributions.

I've included @mindsolve as a reviewer since he was also recently working on the C# implementation.

**Comment by ldilley** [comment]
I think that we can fix the AppVeyor build for this PR by specifying `Visual Studio 2022` as the `image` in `appveyor.yml`. VS 2022 definitely supports C# 8.0 and .NET Core >=3.0/.NET >=5 which features nullable reference types and native JSON (no need for external dependencies which is what @mindsolve wanted to avoid). Are you fine with incrementing our target framework for this purpose, @mindsolve? The disadvantage is that some users will undoubtedly be running older versions of .NET and updating MineStat will break for these users. Of course, these users can continue running older versions of MineStat as a workaround.

**Comment by Ajoro** [comment]
> Apologies for my recent collision (PR #107) with your PowerShell changes in this PR, @Ajoro. I did not realize that this PR also touched those files when the notification came through in Discord (I only saw the C# portion). We can figure out how to merge your work. At any rate, thank you again for the incredible contributions.
> 
> I've included @mindsolve as a reviewer since he was also recently working on the C# implementation.

Added your work to this PR. No problem nice work. (Don't know if that's how you are supposed to do it)

I'm working on alternativ without `using System.Text.Json;` and instead using:
`using System.Runtime.Serialization.Json;
using System.Xml.Linq;
using System.Xml.XPath;`
If you want to keep the older framework.

File: CSharp/MineStat/MineStat.cs [hunk 1]
```diff
@@ -154,43 +165,185 @@ public MineStat(string address, ushort port, int timeout = DefaultTimeout, SlpPr
...
+        RequestWrapper(RequestWithJsonProtocol);
+    }
+    
+    /// <summary>
+    /// Function for stripping all formatting codes from a motd.
+    /// </summary>
+    /// <returns>string with the stripped motd</returns>
+    static private string strip_motd_formatting(string? rawmotd)
+    {
+      return Regex.Replace(rawmotd, @"[^\u0000-\u007F]+[a-zA-Z0-9]", string.Empty);
```
Comment (mindsolve) [inline CSharp/MineStat/MineStat.cs:218]: Why are you searching for all non-ASCII-characters? The string returned by the server is UTF8 (or UTF16) encoded and can contain any of those characters, e.g. Chinese or language-specifics like the German umlauts (äöü), which will trigger this regex. Did you find special cases where other dividiers then `§` were used?
grafik

**Review by mindsolve — CHANGES_REQUESTED** [review]
Hi @Ajoro! 😃

Thanks a lot for the great additions and improvements to the C# and PowerShell versions!

**Regarding the C# version and JSON handling:**
My idea was to keep the different flavors as portable and "single file" as possible. That is why I used the (admittedly quite horrible) `System.Runtime.Serialization.Json` approach, as it was the only fully portable and dependency-free way of having JSON support, while maintaining backwards compatibility with .Net Framework 4.6.
One of the biggest problems with that approach was that I never got serialization to work, which is beautifully resolved by switching to the "new" `System.Text.Json` module.

If we set aside the "dependency-free" idea of mine, we would only have to change our requirements from .NET Framework 4.6 to  4.6.**1**!

[Quote](https://docs.microsoft.com/en-us/dotnet/standard/serialization/system-text-json-overview?pivots=dotnet-6-0#how-to-get-the-library):
> [`System.Text.Json`] is built-in as part of the shared framework for .NET Core 3.0 and later versions. [...]
> For framework versions earlier than .NET Core 3.0, install the [System.Text.Json](https://www.nuget.org/packages/System.Text.Json) NuGet package.
> The package supports:
>    - .NET Standard 2.0 and later

[According to the documentation, .NET Framework 4.6.1 implements .NET Standard 2.0](https://dotnet.microsoft.com/en-us/platform/dotnet-standard#versions), meaning we could add a new dependency to the NuGet package `System.Text.Json` for .NET <= Core 3.0, and then we simply target to .NET Standard 2.0.

What do you think of that, @Ajoro and @ldilley?

File: CSharp/MineStat/MineStat.cs (hunk 1 above)
Comment (Ajoro) [reply CSharp/MineStat/MineStat.cs]: No, it's just an error on my part. Just matching § with `@"\u00A7+[a-zA-Z0-9]"` works!

**Comment by Ajoro** [comment]
> **Regarding the C# version and JSON handling:** My idea was to keep the different flavors as portable and "single file" as possible. That is why I used the (admittedly quite horrible) `System.Runtime.Serialization.Json` approach, as it was the only fully portable and dependency-free way of having JSON support, while maintaining backwards compatibility with .Net Framework 4.6. One of the biggest problems with that approach was that I never got serialization to work, which is beautifully resolved by switching to the "new" `System.Text.Json` module.
> 
> If we set aside the "dependency-free" idea of mine, we would only have to change our requirements from .NET Framework 4.6 to 4.6.**1**!
> 
> [Quote](https://docs.microsoft.com/en-us/dotnet/standard/serialization/system-text-json-overview?pivots=dotnet-6-0#how-to-get-the-library):
> 
> > [`System.Text.Json`] is built-in as part of the shared framework for .NET Core 3.0 and later versions. [...]
> > For framework versions earlier than .NET Core 3.0, install the [System.Text.Json](https://www.nuget.org/packages/System.Text.Json) NuGet package.
> > The package supports:
> > 
> > * .NET Standard 2.0 and later
> 
> [According to the documentation, .NET Framework 4.6.1 implements .NET Standard 2.0](https://dotnet.microsoft.com/en-us/platform/dotnet-standard#versions), meaning we could add a new dependency to the NuGet package `System.Text.Json` for .NET <= Core 3.0, and then we simply target to .NET Standard 2.0.
> 
> What do you think of that, @Ajoro and @ldilley?

That sounds like not that bad of a plan!

Here is where i'm at if we don't edit the targets:
A function that takes the XElement from the descriptionelement  and stripps the format.
```
    Stripped_Motd = strip_motd_formatting(descriptionElement);

    static private string strip_motd_formatting(XElement rawmotd){
      if(rawmotd.FirstAttribute.Value == "string")
        return strip_motd_formatting(rawmotd.FirstNode.ToString());
      var stripped_motd = rawmotd.Element("text")?.Value;
      if (rawmotd.Elements("extra").Any())
      {
        var json_data = rawmotd.Element("extra").Elements();
        foreach (var item in json_data)
        {
          stripped_motd += item.Element("text")?.Value;
        }
      }
      return strip_motd_formatting(stripped_motd);
    }
```
The only problem is that I don't know how to return the motd from XElement into Json. If I use `descriptionElement.ToString()` it  prints the XML but not how the input looked from the beginning. 

In the version with `System.Text.Json` I just serialize the description element: 
`Motd = JsonSerializer.Serialize(payload.GetProperty("description"));`

I was thinking about doing a regex over the raw json payload but I dont think its a good idea... What do you think? :)

**Comment by mindsolve** [comment]
> That sounds like not that bad of a plan!
> 
> Here is where i'm at if we don't edit the targets: A function that takes the XElement from the descriptionelement and stripps the format.
> [...]
> The only problem is that I don't know how to return the motd from XElement into Json. If I use `descriptionElement.ToString()` it prints the XML but not how the input looked from the beginning.
> 
> In the version with `System.Text.Json` I just serialize the description element: `Motd = JsonSerializer.Serialize(payload.GetProperty("description"));`
> 
> I was thinking about doing a regex over the raw json payload but I dont think its a good idea... What do you think? :)

Yeah, that is exactly the problem I was also having trouble with. The old JSON parsing method uses a modified XML parser. ~As such you would have to use a special XML-tree to JSON serializer, which seems to exists, but doesn't work on fragments (e.g. only the MOTD)~
Well, strike that 😄 - I just got it to work. You have to convert the "XML" fragment to a XML document, and then feed it into a JsonWriter. I can't believe that worked, I have tried so many things before - you must be a good influence 😉.
Here is my bad diff, you only have to beautify it a bit 😜....

```diff
diff --git a/CSharp/MineStat/MineStat.cs b/CSharp/MineStat/MineStat.cs
--- a/CSharp/MineStat/MineStat.cs	(revision 73da822f2dce4e0085cba2e30c20ab2c06108d49)
+++ b/CSharp/MineStat/MineStat.cs	(date 1657329089232)
@@ -26,6 +26,7 @@
 using System.IO;
 using System.Linq;
 using System.Runtime.Serialization.Json;
+using System.Xml;
 using System.Xml.Linq;
 using System.Xml.XPath;
 
@@ -300,6 +301,21 @@
         {
           Motd = descriptionElement.Value;
         }
+        else if (descriptionElement != null && descriptionElement.Attribute(XName.Get("type"))?.Value == "object")
+        {
+          // JsonWriter needs a XmlDocument with the root element "root"
+          XmlDocument motdJsonDocument = new XmlDocument();
+          descriptionElement.Name = "root";
+          motdJsonDocument.LoadXml(descriptionElement.ToString());
+          MemoryStream tempMotdJsonStream = new MemoryStream();
+          using (XmlWriter jsonWriter = JsonReaderWriterFactory.CreateJsonWriter(tempMotdJsonStream))
+          {
+            motdJsonDocument.WriteTo(jsonWriter);
+          }
+
+          Motd = Encoding.UTF8.GetString(tempMotdJsonStream.ToArray());
+          //TODO: Generate stripped MOTD from this one
+        }
         else if (root.XPathSelectElement("//description/text") != null)
         {
           Motd = root.XPathSelectElement("//description/text")?.Value;
```

**Comment by ldilley** [comment]
That was a nice find regarding JSON and .NET Framework 4.6.1, @mindsolve. It seems like a clever solution. I would support either path forward (using .NET Framework 4.6.1 or .NET Core >=3.0/.NET >=5). Also, good job figuring out the `XmlDocument` to `JsonWriter` conversion. Thank you for implementing the fixes as well, @Ajoro. It seems that AppVeyor is happy again. :)

The remaining recommendations I have are:
1. Updating the `frameworkAssembly` in `MineStat.nuspec` to reflect `net461`. Additionally, the `TargetFramework` and `TargetFrameworkVersion` in `MineStat.csproj` should be updated accordingly (to `net461` and `v4.6.1`).
2. Since we are including support for a new protocol (Bedrock), I believe this would constitute a major version increase (3.0.0?). Along with incrementing the version in `MineStat.cs`, do not forget to also update the versions at the bottom of `AssemblyInfo.cs`.
3. Feel free to include your username and/or name with the year (2022) in the copyright section contained in the header of `MineStat.cs`, @Ajoro. Lastly, add the same information to the `AssemblyCopyright` line in `AssemblyInfo.cs`.

**Comment by mindsolve** [comment]
>     1. Updating the `frameworkAssembly` in `MineStat.nuspec` to reflect `net461`. Additionally, the `TargetFramework` and `TargetFrameworkVersion` in `MineStat.csproj` should be updated accordingly (to `net461` and `v4.6.1`).
If we keep using the `System.Runtime.Serialization.Json`-way, we would not need to change the targetted framework, as all frameworks after at least .NET Framework 4.0 include support for it.
We would only need to change the target if we were to use the "new" library.

**Comment by Ajoro** [comment]
I think we should keep the the old framework! So I believe we are ready!
Thank you, Mindsolve and Idilley!

**Review by ldilley — APPROVED** [review]

**Review by mindsolve — APPROVED** [review]

**Pull request merged**